### ISSUE TITLE
Macro hygienics don't prevent const cycles

### ISSUE THREAD

**Issue by workingjubilee** [rust-lang/rust#123691]
Consider the following code: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=85373b84b46ca4d49dd9544030fdc081

The most relevant part is this:
```rust
macro_rules! utf16 {
    ($str:expr) => {{
        const UTF8: &str = $str;
        const UTF16_LEN: usize = utf16_len(UTF8);
        const UTF16: [u16; UTF16_LEN] = to_utf16(UTF8);
        &UTF16
    }};
}

fn main() {
    const UTF8: &str = "blargh";
    const UTF16: &[u16] = { utf16!(UTF8) };
}
```

I expected macro hygienics to prevent the const cycle somehow.
Instead it fails to build:
```
error[E0391]: cycle detected when simplifying constant for the type system `main::UTF16::UTF8`
  --> src/main.rs:3:9
   |
3  |         const UTF8: &str = $str;
   |         ^^^^^^^^^^^^^^^^
...
12 |     const UTF16: &[u16] = { utf16!(UTF8) };
   |                             ------------ in this macro invocation
   |
note: ...which requires const-evaluating + checking `main::UTF16::UTF8`...
  --> src/main.rs:12:36
   |
12 |     const UTF16: &[u16] = { utf16!(UTF8) };
   |                                    ^^^^
   = note: ...which again requires simplifying constant for the type system `main::UTF16::UTF8`, completing the cycle
note: cycle used when const-evaluating + checking `main::UTF16::UTF16_LEN`
  --> src/main.rs:4:44
   |
4  |         const UTF16_LEN: usize = utf16_len(UTF8);
   |                                            ^^^^
...
12 |     const UTF16: &[u16] = { utf16!(UTF8) };
   |                             ------------ in this macro invocation
   = note: see https://rustc-dev-guide.rust-lang.org/overview.html#queries and https://rustc-dev-guide.rust-lang.org/query.html for more information
   = note: this error originates in the macro `utf16` (in Nightly builds, run with -Z macro-backtrace for more info)
```

The following diff fixes it:
```diff
 macro_rules! utf16 {
     ($str:expr) => {{
-        const UTF8: &str = $str;
-        const UTF16_LEN: usize = utf16_len(UTF8);
-        const UTF16: [u16; UTF16_LEN] = to_utf16(UTF8);
-        &UTF16
+        const _UTF8: &str = $str;
+        const _UTF16_LEN: usize = utf16_len(_UTF8);
+        const _UTF16: [u16; _UTF16_LEN] = to_utf16(_UTF8);
+         &_UTF16
     }};
 }
```

### Meta
`rustc --version --verbose`:
```
rustc 1.79.0-nightly (ab5bda1aa 2024-04-08)
binary: rustc
commit-hash: ab5bda1aa70f707014e2e691e43bc37a8819252a
commit-date: 2024-04-08
host: x86_64-unknown-linux-gnu
release: 1.79.0-nightly
LLVM version: 18.1.3
```

**Comment by ChrisDenton** [comment]
A more minimal repo would be:

```rust
macro_rules! cycle {
    ($expr:expr) => {{
        const UTF8: &str = $expr;
    }}
}
#[allow(unused)]
fn main() {
    const UTF8: &str = "hello";
    cycle!(UTF8);
}
```

**Comment by Urgau** [comment]
Not sure if it's relevant to this issue but the same error happens without the `macro_rules!`:
```rust
#[allow(dead_code)]
fn main() {
    const UTF8: &str = "hello";
    {
        const UTF8: &str = UTF8;
    }
}
```

**Comment by fmease** [comment]
This works as intended. [typeing explainer...]

**Comment by fmease** [comment]
> A more minimal repo would be:

Even more minimal:

```rs
macro_rules! cycle {
    ($expr:expr) => { const X: () = $expr; }
}

fn main() { cycle!(X); } //~ ERROR cycle detected
```

---

`macro_rules!`-macros aka decl macros 1.2 are only *partially hygienic* because only local variables follow macro hygiene but not *items*. When you define constant `X` via a 1.2 macro, you can access it from other syntax contexts. That's what happens in the example above. Another example would be

```rs
macro_rules! def { () => { const X: () = (); } }

fn main() {
    def!();
    let _ = X; // successfully resolves `X`
}
```

Decl macros 2.0 (`#![feature(decl_macro)]`) on the other hand are *fully* hygienic: Both locals and items are hygienic. For example, the following code succeeds compilation:

```rs
#![feature(decl_macro)]

macro cycle {
    ($expr:expr) => { const X: () = $expr; }
}

fn main() {
    const X: () = ();
    cycle!(X);
}
```

**Comment by fmease** [comment]
TL;DR: Your example from the issue description passes compilation if you use decl macros 2.0 instead of macros 1.2: [Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=386550f4026a08598b1ad9101aace2c5).

**Comment by fmease** [comment]
Closing as works as intended :)

**Comment by workingjubilee** [comment]
I had a feeling this might get a "working as intended" but I'm glad to hear it gets fixed by macros 2.0!
